blockchain/v2/scheduler.go

```diff
@@ -599,11 +628,13 @@ func (sc *scheduler) handleRemovePeer(event bcRemovePeer) (Event, error) {
 	if sc.allBlocksProcessed() {
 		return scFinishedEv{reason: "removed peer"}, nil
 	}
-	return noOp, nil
+
+	// Return scPeerError so the peer (and all associated blocks) is removed from
+	// the processor.
+	return scPeerError{peerID: event.peerID, reason: errors.New("peer was stopped")}, nil
```

Comment:
Interesting, this isn't obvious to me. Does this mean that whenever we remove a peer, we also discard all the blocks we've fetched from it? What causes the peer to be removed? For example, if we've fetched 100 blocks from a peer, and the peer then disconnects (e.g. due to operator shutdown), we just throw away all of those 100 blocks?

Reply:
That's how it was implemented. NOTE we only throw away the blocks "in flight" - blocks we've downloaded, but haven't yet processed.

Reply:
Wouldn't it be better to keep those blocks? There may be cases where we should discard the blocks (e.g. if a peer supplied an invalid block), but if the peer simply disconnected shouldn't we at least attempt to apply those blocks instead?

That said, it seems like the current code intends to discard those blocks, so maybe that's fine for now. It's definitely safer, just possibly wasteful.

Reply:
Yeah, let's lean on the safer side for now. Created: https://github.com/tendermint/tendermint/issues/5532